marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/BuildConstraintsSpec.hs

```diff
@@ -223,3 +248,99 @@ withdrawSpec = Hspec.describe "buildWithdrawConstraints" do
           }
 
     pure $ actual `shouldBe` expected
+
+buildApplyInputsConstraintsSpec :: Spec
+buildApplyInputsConstraintsSpec =
+  Hspec.describe "buildApplyInputsConstraints" do
+    -- TODO: Break up this preface to utility functions when additional properties are tested.
+    let
+      -- System start and era history are reusable across tests.
+      systemStart = SystemStart $ posixSecondsToUTCTime 0  -- Without loss of generality.
+      eraParams = EraParams
+        { eraEpochSize = 1
+        , eraSlotLength = mkSlotLength 1
+        , eraSafeZone = UnsafeIndefiniteSafeZone
+        }
+      oneSecondBound i = Bound
+        { boundTime = RelativeTime $ fromInteger i
+        , boundSlot = fromInteger i
+        , boundEpoch = fromInteger i
+        }
+      oneSecondEraSummary i = EraSummary
+        { eraStart = oneSecondBound i
+        , eraEnd = EraEnd $ oneSecondBound $ i + 1
+        , eraParams
+        }
+      unboundedEraSummary i = EraSummary
+        { eraStart = oneSecondBound i
+        , eraEnd = EraUnbounded
+        , eraParams
+        }
+      eraHistory =
+        EraHistory CardanoMode
+          . mkInterpreter
+          . summaryWithExactly
+          $ Exactly
+          $  K (oneSecondEraSummary 0) -- Byron lasted 1 second
+          :* K (oneSecondEraSummary 1) -- Shelley lasted 1 second
+          :* K (oneSecondEraSummary 2) -- Allegra lasted 1 second
+          :* K (oneSecondEraSummary 3) -- Mary lasted 1 second
+          :* K (oneSecondEraSummary 4) -- Alonzo lasted 1 second
+          :* K (unboundedEraSummary 5) -- Babbage never ends
+          :* Nil
+      genTipSlot = chooseInteger (9, 20) -- Make sure the tip is in the Babbage era.
+      genMinTime = oneof
+                   [
+                     (1000 *) <$> chooseInteger (6, 20)  -- Even seconds, so there is a chance of collision.
+                   , chooseInteger (6000, 20000)         -- Milliseconds, so there is rounding off.
+                   ]
+      genTimeout = oneof
+                   [
+                     (1000 *) <$> chooseInteger (6, 20)  -- Even seconds, so there is a chance of collision.
+                   , chooseInteger (6000, 20000)         -- Milliseconds, so there is rounding off.
+                   ]
+    Hspec.QuickCheck.prop "valid slot interval for timed-out contract" \assets utxo address marloweParams-> do
+      -- The choice intervals for the tip, minimum time, and timeout overlap, so every ordering will occur.
+      tipSlot' <- genTipSlot
+      minTime <- genMinTime
+      timeout <- genTimeout
+      timeout' <- genTimeout
+      let
+        contract = Semantics.When [] (POSIXTime timeout) Semantics.Close
+        contract' = Semantics.When [] (POSIXTime timeout) $ Semantics.When [] (POSIXTime timeout') Semantics.Close
```

Comment:
@palas and @jhbertra, should we add deeper test contracts here?
- *Pro:* it's a more comprehensive of timeout traversal.
- *Con:* this is more controlled that with a general arbitrary contract, where we'd more-or-less have to recode the "subsequent timeout" logic that we are testing. If we go with a general arbitrary contract, then I'd like to have a completely different timeout analysis in the test code, as compared to the code being tested.

Reply:
I don't think an arbitrary contract makes sense here - but we could potentially generate a contract using a custom generator (and `forallShrink`) which generates contracts appropriate for the property. This could also be tackled separately though.